sdk/cosmosdb/cosmos/src/request/CosmosDiagnostics.ts

```diff
@@ -0,0 +1,212 @@
+  public constructor() {
+    this.requestStartTimeUTC = getCurrentTimestamp();
+    this.retryStartTimeUTC = this.requestStartTimeUTC;
+  }
+
+  public recordRequestPayload(payload: string): void {
+    this.requestPayloadLength = payload.length;
+  }
+
+  public recordResponseStats(payload: string, headers: CosmosHeaders): void {
+    this.responsePayloadLength = typeof payload === "string" ? payload.length : 0;
+    this.headers = { ...this.headers, ...headers };
+  }
+
+  public recordFailedAttempt(statusCode: string): void {
+    const attempt: FailedRequestAttempt = {
+      id: v4(),
+      attemptNumber: this.retryAttempNumber,
+      startTimeUTC: this.retryStartTimeUTC,
+      endTimeUTC: getCurrentTimestamp(),
+      statusCode,
+    };
+    this.retryStartTimeUTC = getCurrentTimestamp();
+    this.retryAttempNumber++;
+    this.failedAttempts.push(attempt);
+  }
+
+  public recordMetaDataLookup(diagnostics: CosmosDiagnostics, metaDataType: MetadataType): void {
+    const metaDataRequest = {
+      startTimeUTC: diagnostics.clientSideRequestStatistics.requestStartTimeUTC,
+      endTimeUTC: diagnostics.clientSideRequestStatistics.requestEndTimeUTC,
+      activityId: diagnostics.clientSideRequestStatistics.activityId,
+      metaDataType,
+      id: v4(),
+    };
+    this.metadataLookups.push(metaDataRequest);
+  }
+
+  //   public recordSerializationEvent() {}
+
+  //   public recordResponse() {}
+
+  public mergeDiagnostics(diagnostics: CosmosDiagnostics): void {
+    diagnostics.clientSideRequestStatistics.locationEndpointsContacted.forEach((endpoint) =>
+      this.locationEndpointsContacted.add(endpoint)
+    );
+    diagnostics.clientSideRequestStatistics.metadataDiagnostics.metadataLookups.forEach((lookup) =>
+      this.metadataLookups.push(lookup)
+    );
+    diagnostics.clientSideRequestStatistics.retryDiagnostics.failedAttempts.forEach((lookup) =>
+      this.failedAttempts.push(lookup)
+    );
+  }
+
+  public recordEndpointContactEvent(location: string): void {
+    this.locationEndpointsContacted.add(location);
+  }
+
+  public getDiagnostics(): CosmosDiagnostics {
+    this.recordSessionEnd();
+    return {
+      id: v4(),
+
+      clientSideRequestStatistics: {
+        activityId: this.getActivityId(),
+        requestStartTimeUTC: this.requestStartTimeUTC,
+        requestEndTimeUTC: this.requestEndTimeUTC,
+        locationEndpointsContacted: [...this.locationEndpointsContacted],
+        metadataDiagnostics: {
+          metadataLookups: [...this.metadataLookups],
+        },
+        retryDiagnostics: {
+          failedAttempts: [...this.failedAttempts],
+        },
+        requestPayloadLength: this.requestPayloadLength,
+        responsePayloadLength: this.responsePayloadLength,
+      },
+    };
+  }
+
+  private recordSessionEnd() {
+    this.requestEndTimeUTC = getCurrentTimestamp();
+  }
+
+  private getActivityId(): string {
+    return this.headers[Constants.HttpHeaders.ActivityId] as string;
+  }
+}
```

Comment:
Should not we need other public methods to align with other SDKs? getRegionsContacted and getContactedRegionNames

Reply:
Added.